File: 20-actors/fuchi/methods/test_analyze.py
#!/usr/bin/env python3
"""End-to-end tests for 扶持 (fuchi) analyze.py over the :representative seed.

Standalone-runnable: python3 test_analyze.py
"""
from __future__ import annotations

import sys

import analyze


def _rows():
    return {r["did"].split(":")[-1]: r for r in analyze.run()["rows"]}


def test_abel_auto_accepts():
    r = _rows()["abel"]
    assert r["route"] == ":auto" and r["outcome"] == "accepted"


def test_seth_goes_to_sbt_vote_and_passes():
    r = _rows()["seth"]
    assert r["route"].startswith(":sbt-vote") and r["outcome"] == "accepted"


def test_eve_escalates_to_council_lv7():
    r = _rows()["eve"]
    assert r["route"] == ":council-lv7" and r["outcome"] == "pending"


def test_cain_is_refused_by_rider():
    r = _rows()["cain"]
    assert r["route"] == ":refused" and r["outcome"] == "refused"


def test_noah_outreach_zero_share():
    r = _rows()["noah"]
    assert r["covenant"] == "outreach" and r["share"] == 0.0


def test_seth_in_kind_coverage_below_one():
    # seth carries an external liquidity residual → in-kind < 100%
    assert _rows()["seth"]["in_kind"] < 1.0


def test_every_derived_alloc_has_zero_cash():
    for d in analyze.run()["derived"]:
        assert d[":alloc/cash-usd-micros"] == 0
        assert d[":alloc/server-held-key"] is False


def test_refused_alloc_not_in_derived():
    dids = {d[":alloc/maintainer"].split(":")[-1] for d in analyze.run()["derived"]}
    assert "cain" not in dids


# ── R1 a/b/c end-to-end ─────────────────────────────────────────────────────
def test_seth_vote_finalized_after_timelock():
    # the route string carries the ✓ only when the 48h window has closed (R1b)
    assert "✓" in _rows()["seth"]["route"]


def test_provisioning_intents_emitted():
    res = analyze.run()
    assert res["intents"], "expected provisioning intents (R1a)"
    # every intent is a dry-run (published false, cash 0, keyless)
    for i in res["intents"]:
        assert i.published is False and i.cash_usd_micros == 0 and i.server_held_key is False


def test_seth_liquidity_intent_is_member_principal():
    res = analyze.run()
    liq = [i for i in res["intents"]
           if i.alloc_id.endswith("seth") and i.rail_kind == "liquidity-warifu"]
    assert liq and liq[0].member_principal is True


def test_toritate_ledger_is_cashless_no_liquidity():
    res = analyze.run()
    assert res["ledger"], "expected toritate ledger entries (R1c)"
    for e in res["ledger"]:
        assert e.cash_usd_micros == 0
    # liquidity is never booked as income → seth has no ledger entry > vocation/subsistence/care only
    cats = {e.category for e in res["ledger"]}
    assert cats <= {"subsistence-flow", "vocation-flow", "care-flow", "liberation-flow", "grant"}


def test_flow_graph_has_publicfund_source():
    res = analyze.run()
    assert any(f.flow_class == "publicfund-to-fuchi" for f in res["flows"])


# ── R1(d) coupling ──────────────────────────────────────────────────────────
def _coupling():
    return {c["earmark"].cohort_id: c for c in analyze.run()["coupling"]}


def test_sanae_cohort_funded_and_admissible():
    c = _coupling()["cohort-sanae-2026"]
    assert c["earmark"].funded is True and c["gate"]["admissible"] is True
    # 10% split exact
    assert c["earmark"].tithe_usd_micros == 6_000_000_000
    assert c["earmark"].earmark_usd_micros_yr == 54_000_000_000


def test_hataori_cohort_unfunded_refused():
    c = _coupling()["cohort-hataori-2026"]
    assert c["earmark"].funded is False and c["gate"]["admissible"] is False
    assert "G2" in c["gate"]["reason"]


def _run():
    fns = [v for k, v in sorted(globals().items()) if k.startswith("test_") and callable(v)]
    for fn in fns:
        fn()
    print(f"test_analyze.py: {len(fns)} passed")
    return 0


if __name__ == "__main__":
    sys.exit(_run())

File: 20-actors/fuchi/methods/test_analyze.clj
;; test_analyze.clj — 扶持 analyze: end-to-end allocation pipeline over the :representative seed,
;; parity with analyze.py run(). Run via `bb test:fuchi`. ADR-2606142300.
(ns fuchi.methods.test-analyze
  (:require [clojure.test :refer [deftest is testing run-tests]]
            [clojure.string :as str]
            [fuchi.methods.analyze :as a]))

(def ^:private res (a/run))

(defn- row [did] (first (filter #(str/ends-with? (str (:did %)) did) (:rows res))))

(deftest pipeline-routing-outcomes
  (testing "each seed maintainer routes + resolves exactly as analyze.py run() (golden)"
    (is (= 5 (count (:rows res))))
    (is (= ":auto"        (:route (row "abel"))))   (is (= "accepted" (:outcome (row "abel"))))
    (is (= "accepted"     (:outcome (row "seth"))))                          ; via sbt-vote
    (is (str/starts-with? (:route (row "seth")) ":sbt-vote 5-1/48h✓"))       ; real 1 SBT=1 vote tally
    (is (= ":council-lv7" (:route (row "eve"))))    (is (= "pending"  (:outcome (row "eve"))))
    (is (= "accepted"     (:outcome (row "noah"))))                          ; outreach, auto
    (is (= ":refused"     (:route (row "cain"))))   (is (= "refused"  (:outcome (row "cain"))))))

(deftest pipeline-projection-counts
  (testing "provisioning / ledger / flow / derived counts match analyze.py run()"
    (is (= 14 (count (:intents res))))
    (is (= 13 (count (:ledger res))))
    (is (= 32 (count (:flows res))))
    (is (= 4  (count (:derived res))))))           ; abel+seth+eve+noah booked; cain refused

(deftest cash-zero-everywhere
  (testing "cash≡0 holds across every projection (the defining invariant)"
    (is (every? #(= 0 (:cash-usd-micros %)) (:intents res)))
    (is (every? #(= 0 (:cash-usd-micros %)) (:ledger res)))
    (is (every? #(false? (:server-held-key %)) (:intents res)))
    (is (every? #(false? (:published %)) (:intents res)))         ; G10 dry-run
    (is (every? #(= 0 (:alloc/cash-usd-micros %)) (:derived res)))))

(deftest displacement-coupling-G2
  (testing "Displacement-Dividend coupling: funded cohort admissible, unfunded refused (golden)"
    (let [by (into {} (map (juxt #(str (get-in % [:earmark :cohort-id])) identity) (:coupling res)))]
      (is (= 2 (count (:coupling res))))
      (is (true?  (get-in (by "cohort-sanae-2026") [:gate :admissible])))   ; funded
      (is (false? (get-in (by "cohort-hataori-2026") [:gate :admissible])))  ; unfunded → G2 refuse
      (is (= 6000000000 (get-in (by "cohort-sanae-2026") [:earmark :tithe-usd-micros]))))))  ; 10% of 60e9

(deftest live-legs-and-scorecard
  (testing "all four R2 live legs admissible; scorecard renders"
    (is (= #{"provision" "vote" "book" "couple"} (set (map :leg (:live-status res)))))
    (is (every? :admissible (:live-status res)))
    (let [sc (a/scorecard res)]
      (is (str/includes? sc "扶持"))
      (is (str/includes? sc "accepted: 3"))                       ; abel + seth + noah
      (is (str/includes? sc "refused: 1")))))                     ; cain

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (run-tests 'fuchi.methods.test-analyze)]
    (System/exit (if (pos? (+ fail error)) 1 0))))
